state/pruner.go

```diff
@@ -115,23 +115,15 @@ func (p *Pruner) SetApplicationRetainHeight(height int64) error {
 	if !p.checkHeightBound(height) {
 		return ErrInvalidHeightValue
 	}
-	currentAppRetainHeight, err := p.stateStore.GetApplicationRetainHeight()
-	if err != nil {
-		if !errors.Is(err, ErrKeyNotFound) {
-			return err
-		}
-		currentAppRetainHeight = height
+	currentMinRetainHeight := p.findMinRetainHeight()
+	if height < currentMinRetainHeight {
+		return ErrPrunerCannotLowerRetainHeight
```

Comment:
Not sure if this logic returns the desirable outcome. For example, let's assume the following scenario:

| DCRetainHeight | AppRetainHeight | Height |
|----------------|-----------------|--------|
|       98       |       100       |   99   |

This logic would return a `currentMinRetainHeight = 98` and since `height (99) > currentMinRetainHeight (99)` this would set the `AppRetainHeight` to `99` but it was already at `100`, so after executing this logic you would end up with:

| DCRetainHeight | AppRetainHeight | Height |
|----------------|-----------------|--------|
|       98       |        99       |   99   |

I'm thinking that the logic in `p.findMinRetainHeight()` shouldn't be used in the logic to `Set` the retain heights, because it mixes the DC and the App retain height which might make the things more complicated. Maybe we just need to check here if the current AppRetainHeight is lower or not and set it accordingly. 

Then in the pruner logic to check between the AppRetainHeight and DCRetainHeight whichever is lower then we use the lower one (the one returned by `p.findMinRetainHeight`

Reply:
Can we perhaps introduce a comprehensive set of tests for the desired behaviour?